From 5d685a69506c5a886aadebf1643683cbde8c5f1f Mon Sep 17 00:00:00 2001
From: Rob Herring <r.herring@freescale.com>
Date: Tue, 3 Mar 2009 10:46:28 -0600
Subject: [PATCH] ENGR00108843 ipu3: support bigger hsync widths

The hsync pulse width was based on the DI clock and would easily overflow
the 9-bit field. This problem was seen with standard VESA timings. Changing
the waveform counter to be based on the pixel clock instead.

Also, add some checks for overflowing the register fields of the DI counters.

Signed-off-by: Rob Herring <r.herring@freescale.com>
---
 drivers/mxc/ipu3/ipu_disp.c |   10 ++++++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/ipu3/ipu_disp.c b/drivers/mxc/ipu3/ipu_disp.c
index d0eafa2..979bfed 100644
--- a/drivers/mxc/ipu3/ipu_disp.c
+++ b/drivers/mxc/ipu3/ipu_disp.c
@@ -137,6 +137,12 @@ static void _ipu_di_sync_config(int di, int wave_gen,
 				int cnt_polarity_trigger_src,
 				int cnt_up, int cnt_down)
 {
+	if ((run_count >= 0x1000) || (offset_count >= 0x1000) || (repeat_count >= 0x1000) ||
+		(cnt_up >= 0x400) || (cnt_down >= 0x400)) {
+		dev_err(g_ipu_dev, "DI%d counters out of range.\n", di);
+		return;
+	}
+
 	u32 reg;
 	reg = (run_count << 19) | (++run_src << 16) |
 	    (offset_count << 3) | ++offset_src;
@@ -879,8 +885,8 @@ int32_t ipu_init_sync_panel(int disp, uint32_t pixel_clk,
 		/* Setup external (delayed) HSYNC waveform */
 		_ipu_di_sync_config(disp, DI_SYNC_HSYNC, h_total - 1,
 				    DI_SYNC_CLK, div * v_to_h_sync, DI_SYNC_CLK,
-				    0, DI_SYNC_NONE, 0, DI_SYNC_NONE,
-				    DI_SYNC_NONE, 0, div * h_sync_width * 2);
+				    0, DI_SYNC_NONE, 1, DI_SYNC_NONE,
+				    DI_SYNC_CLK, 0, h_sync_width * 2);
 		/* Setup VSYNC waveform */
 		vsync_cnt = DI_SYNC_VSYNC;
 		_ipu_di_sync_config(disp, DI_SYNC_VSYNC, v_total - 1,
-- 
1.5.4.4

